src/Passes/Source/Rules/OperandPrototype.cpp

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "Rules/OperandPrototype.hpp"
+
+namespace microsoft
+{
+namespace quantum
+{
+
+    IOperandPrototype::~IOperandPrototype() = default;
+    bool IOperandPrototype::matchChildren(Value* value, Captures& captures) const
+    {
+        auto user = llvm::dyn_cast<llvm::User>(value);
+        if (!children_.empty())
+        {
+            if (user == nullptr)
+            {
+                return false;
+            }
+
+            if (user->getNumOperands() != children_.size())
+            {
+                return false;
+            }
+
+            uint64_t i = 0;
+            while (i < children_.size())
+            {
+                auto v = user->getOperand(static_cast<uint32_t>(i));
+                if (!children_[i]->match(v, captures))
+                {
+                    return false;
+                }
+                ++i;
+            }
+
+            return true;
+        }
+
+        // TODO(QAT-private-issue-33): value may be other type than llvm::User. Check other relevant types
+        // and deal with it.
+
+        return true;
+    }
+
+    void IOperandPrototype::addChild(Child const& child)
+    {
+        children_.push_back(child);
+    }
+
+    void IOperandPrototype::enableCapture(std::string capture_name)
+    {
+        capture_name_ = std::move(capture_name);
+    }
+
+    bool IOperandPrototype::fail(Value* /*value*/, Captures& /*captures*/) const
+    {
+        return false;
+    }
+
+    bool IOperandPrototype::success(Value* value, Captures& captures) const
+    {
+        capture(value, captures);
+
+        auto ret = matchChildren(value, captures);
+        if (!ret)
+        {
+            uncapture(value, captures);
+        }
+        return ret;
+    }
+
+    void IOperandPrototype::capture(Value* value, Captures& captures) const
+    {
+        if (!capture_name_.empty())
+        {
+            captures[capture_name_] = value;
+        }
+    }
+
+    void IOperandPrototype::uncapture(Value* /*value*/, Captures& captures) const
+    {
+        if (!capture_name_.empty())
+        {
+            captures.erase(captures.find(capture_name_));
+        }
+    }
```

Comment:
Important:

What will happen if `capture()` is called one time, but `uncapture()` is called 2 times? I failed to find the answer [here](https://cplusplus.com/reference/unordered_map/unordered_map/erase/) or [here](https://en.cppreference.com/w/cpp/container/unordered_map/erase).

I'm afraid that the undefined behavior will happen and the `captures.erase()` will be able to _corrupt the memory silently_.
Consider asserting or throwing in case if `captures.find()` returns `end()` (during the disbalanced call to `uncapture()`).

Reply:
That is a good point. While this is a potential problem, these functions were intended to be implementation details of the class. Hence they should indeed not be in the protected section but in the private section.

Moreover, while the class is supposed to provide a guarantee that these are called correctly, I think it is a valid point that if ever someone would succeed in call `capture()` once and `uncapture()` twice it would be a difficult bug to debug. 

I've moved the functions to the private section and added a check prior to deleting. It can be debated whether this should be an assert or an exception. I've gone for an exception, but please let me know if you would prefer an assert.